// dark/dist/option.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Camel-case the given `string`
 *
 * @param string - the string to camelcase
 */
function camelcase(string) {
  return string.split('-').reduce((str, word) => str + word[0].toUpperCase() + word.slice(1));
}
/**
 * Initialize a new [[Option]] with the given `flags` and `description`.
 *
 * @param flags
 * @param description
 */


class Option {
  constructor(flags, description) {
    _defineProperty(this, "flags", void 0);

    _defineProperty(this, "required", void 0);

    _defineProperty(this, "optional", void 0);

    _defineProperty(this, "bool", void 0);

    _defineProperty(this, "short", void 0);

    _defineProperty(this, "long", void 0);

    _defineProperty(this, "description", void 0);

    _defineProperty(this, "defaultValue", void 0);

    this.flags = flags;
    this.required = flags.includes('<');
    this.optional = flags.includes('[');
    this.bool = !flags.includes('-no-');
    const parsedFlags = flags.split(/[ ,|]+/);

    if (parsedFlags.length > 1 && !/^[[<]/.test(parsedFlags[1])) {
      this.short = parsedFlags.shift() || '';
    }

    this.long = parsedFlags.shift() || '';
    this.description = description || '';
  }
  /**
   * Return option name.
   */


  name() {
    return this.long.replace('--', '').replace('no-', '');
  }
  /**
   * Return option name, in a camelcase format that can be used
   * as a object attribute key.
   */


  attributeName() {
    return camelcase(this.name());
  }
  /**
   * Check if `arg` matches the short or long flag. Internally used to check
   * if an arg belongs to an [[Option]]
   */


  is(arg) {
    return this.short === arg || this.long === arg;
  }
  /**
   * Returns number of args that are expected by the option.
   * Can only be 0 or 1.
   */


  arity() {
    return this.required || this.optional ? 1 : 0;
  }

}

exports.default = Option;
